server/src/main/java/org/elasticsearch/index/mapper/ParseContext.java

```diff
@@ -164,7 +164,7 @@ public ObjectMapper getObjectMapper(String name) {
         }
 
         @Override
-        public Mapper.TypeParser.ParserContext dynamicTemplateParserContext(DateFormatter dateFormatter) {
+        public Mapper.TypeParser.ParserContext.DynamicTemplateParserContext dynamicTemplateParserContext(DateFormatter dateFormatter) {
```

Comment:
For conciseness, I would consider renaming this method back to parserContext now that its return type is specific

Reply:
Alternatively, I think the type of this method can stay as just `Mapper.TypeParser.ParserContext`? And then DynamicTemplateParserContext can become a top-level package-private class and we lose some of the crazy nesting :)

Reply:
Counter proposal: shall we get ParserContext out of Mapper.TypeParser completely as a follow-up? it's used for runtime fields too so it is not specific to Mappers, and we could rename it to to something that is less confusing compared to ParseContext that is instead used for documents.

Reply:
> shall we get ParserContext out of Mapper.TypeParser completely as a follow-up?

Strong +1 from me! I think we can still leave this method return type untouched though?

Reply:
Ok good discussion: changing the types makes sure that the wrong context cannot be provided, and helps readers understand that when parsing dynamic templates only the appropriate context can be used.

I guess I see how it can be cleaner and tempting to use the more generic type, but what are the advantages in this case compared to the confusion that it adds in that it is hard to follow and very easy to miss?

Reply:
My argument would be that: we only have a single concrete implementation of ParseContext (and are unlikely to add more) so I don't think we need to worry about returning an incorrect type elsewhere; that the name of the method describes the fact that it's specifically for dynamic templates; in general it's best to return interfaces rather than implementations; and that having four layers of nested class name makes my eyes hurt.

Reply:
the nesting I agree on. My point is that with keeping the type generic we may end up making the same mistake that got us here aka carrying around the wrong context. The name of a method does not enforce anything, while the return type does. I agree on the software engineering guidelines around using interfaces but they also introduce complexity. The important bit is that parsing mappings requires a ParserContext, which is the case. But we can clarify that when we parse as part of dynamic templates, the context that is provided will be specific, we don't need to keep things overly generic everywhere as that brings no benefits?

Reply:
The previous bug was because you had do to two things: 1) get a new parser context from the ParseContext, and 2) convert it to a dynamic context.  This commit merges the two, so that what you get back from 1) is always a dynamic context.  There's no way of getting a parser context from ParseContext that isn't dynamic.  So I don't see how we could make the same mistake again?

Reply:
I feel like it is still a bit delicate if we use the more generic type, we may change that due to refactoring, not too sure. Maybe I am being a bit paranoid but this is still a bit sketchy. Anyways, I will do as you ask.